node/node.go

```diff
@@ -701,3 +708,73 @@ func (node *Node) ShutDown() {
 	fmt.Print(msg)
 	os.Exit(0)
 }
+
+func (node *Node) populateSelfAddresses(epoch *big.Int) {
+	// reset the self addresses
+	node.SelfAddresses = map[string]common.Address{}
+	node.selfAddressEpoch = epoch
+
+	shardID := node.Consensus.ShardID
+	shardState, err := node.Consensus.ChainReader.ReadShardState(epoch)
+	if err != nil {
+		utils.Logger().Error().Err(err).
+			Int64("epoch", epoch.Int64()).
+			Uint32("shard-id", shardID).
+			Msg("[PopulateSelfAddresses] failed to read shard")
+		return
+	}
+
+	committee, err := shardState.FindCommitteeByID(shardID)
+	if err != nil {
+		utils.Logger().Error().Err(err).
+			Int64("epoch", epoch.Int64()).
+			Uint32("shard-id", shardID).
+			Msg("[PopulateSelfAddresses] failed to find shard committee")
+		return
+	}
+
+	for _, blskey := range node.Consensus.PubKey.PublicKey {
+		shardkey := *shard.FromLibBLSPublicKeyUnsafe(blskey)
+		addr, err := committee.AddressForBLSKey(shardkey)
+		blsStr := blskey.SerializeToHexStr()
+		if err != nil {
+			utils.Logger().Error().Err(err).
+				Int64("epoch", epoch.Int64()).
+				Uint32("shard-id", shardID).
+				Str("blskey", blsStr).
+				Msg("[PopulateSelfAddresses] could not find address")
+			return
+		}
+		node.SelfAddresses[blsStr] = *addr
+		utils.Logger().Debug().
+			Int64("epoch", epoch.Int64()).
+			Uint32("shard-id", shardID).
+			Str("bls-key", blsStr).
+			Str("address", common2.MustAddressToBech32(*addr)).
+			Msg("[PopulateSelfAddresses]")
+	}
+}
+
+// GetAddressForBLSKey retrieves the ECDSA address associated with bls key for epoch
+func (node *Node) GetAddressForBLSKey(blskey *bls.PublicKey, epoch *big.Int) common.Address {
+	// populate if first time setting or new epoch
+	if node.selfAddressEpoch == nil || epoch.Cmp(node.selfAddressEpoch) != 0 {
+		node.populateSelfAddresses(epoch)
+	}
+	blsStr := blskey.SerializeToHexStr()
+	addr, ok := node.SelfAddresses[blsStr]
+	if !ok {
+		return common.Address{}
+	}
+	return addr
+}
+
+// GetSelfAddresses retrieves all ECDSA addresses of the bls keys for epoch
+func (node *Node) GetSelfAddresses(epoch *big.Int) map[string]common.Address {
+	// populate if first time setting or new epoch
+	if node.selfAddressEpoch == nil || epoch.Cmp(node.selfAddressEpoch) != 0 {
+		node.populateSelfAddresses(epoch)
```

Comment:
prefer explicit value returned rather than side effect driven. populate could be a plain function that we use to assign to some node.<field>

Reply:
as the map is never nil and the caller is basically using it to update the web hook, I would like to keep it side effect based and not return error.